feat: [M3-7665] - RegionMultiSelect Component

cpathipa opened the pull request:
## Description 📝
This Allows to select multiple regions and renders the list of selected regions

## Changes  🔄
- RemovableSelectionsList - Added an optional prop to render custom label component.


## Preview 📷

| Region Multi select  | Renders selected list   |
| ------- | ------- |
|![image](https://github.com/linode/manager/assets/119517080/6351e96b-a47f-41da-b061-db9084de1625)|![image](https://github.com/linode/manager/assets/119517080/8016f6e2-8dc0-4ce6-ac1c-2eb649504742)|

## How to test 🧪
- Run story book verify the functionality of Region Multi select component.


## As an Author I have considered 🤔

*Check all that apply*

- [ ] 👀 Doing a self review
- [ ] ❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
- [ ] 🤏 Splitting feature into small PRs
- [ ] ➕ Adding a changeset
- [ ] 🧪 Providing/Improving test coverage
- [ ] 🔐 Removing all sensitive information from the code and PR description
- [ ] 🚩 Using a feature flag to protect the release
- [ ] 👣 Providing comprehensive reproduction steps
- [ ] 📑 Providing or updating our documentation
- [ ] 🕛 Scheduling a pair reviewing session
- [ ] 📱 Providing mobile support
- [ ] ♿  Providing accessibility support

jaalah-akamai commented on packages/manager/src/components/RegionSelect/RegionMultiSelect.tsx:
```diff
@@ -0,0 +1,204 @@
+    [accountAvailability, currentCapability, regions]
+  );
+
+  const removeOption = (optionToRemove: RegionSelectOption) => {
```
```suggestion
  const handleRemoveOption = (optionToRemove: RegionSelectOption) => {
```

jaalah-akamai commented on packages/manager/src/components/RegionSelect/RegionMultiSelect.tsx:
```diff
@@ -0,0 +1,204 @@
+      </StyledAutocompleteContainer>
+      {selectedRegions.length > 0 && RenderSelectedRegionsList && (
+        <RenderSelectedRegionsList
+          onRemove={removeOption}
```
```suggestion
          onRemove={handleRemoveOption}
```

jaalah-akamai commented:
The dark mode background colors/shading is a little weird. Do we also want to have a border under `Select All` and `Deselect All` like in the original Autocomplete example?

abailly-akamai commented on packages/manager/src/components/RegionSelect/RegionMultiSelect.stories.tsx:
```diff
@@ -0,0 +1,105 @@
+  selectedRegions,
+}: SelectedRegionsProps) => {
+  const handleRemove = (data: RemovableItem) => {
+    const { id: _id, ...dataWithoutId } = data;
```
This is confusing - why are you doing this?

cpathipa replied:
Added this because the 'RemovableItem' structure consists of an 'id' that is not present in RegionSelectOption. So, I thought it best to eliminate it. However, I now think we don't need it here, so I have removed it.

abailly-akamai commented on packages/manager/src/components/RegionSelect/RegionMultiSelect.stories.tsx:
```diff
@@ -0,0 +1,105 @@
+export const Default: StoryObj<RegionMultiSelectProps> = {
+  render: (args) => {
+    const SelectWrapper = () => {
+      const [selectedRegions, setSelectedRegions] = useState<string[]>([]);
```
I know it's a story, but we can be a little better at naming things here. ex: 
```suggestion
      const [selectedRegionsIds, setSelectedRegionsIds] = useState<Region['id'][]>([]);
```
for better readability

abailly-akamai commented on packages/manager/src/components/RegionSelect/RegionMultiSelect.stories.tsx:
```diff
@@ -0,0 +1,105 @@
+      const [selectedRegions, setSelectedRegions] = useState<string[]>([]);
+
+      const handleSelectionChange = (selectedIds: string[]) => {
+        debugger;
```
remove debugger

abailly-akamai commented on packages/manager/src/components/RegionSelect/RegionMultiSelect.test.tsx:
```diff
@@ -0,0 +1,29 @@
+
+    screen.getByRole('combobox', { name: 'Regions' });
+  });
+});
```
we need better coverage here and test interaction as well as checking a region properly gets added to the list underneath

abailly-akamai commented on packages/manager/src/components/RegionSelect/RegionMultiSelect.tsx:
```diff
@@ -0,0 +1,204 @@
+    return -1;
+  }
+  return 0;
+};
```
![Screenshot 2024-01-22 at 10 23 36](https://github.com/linode/manager/assets/130582365/d3dea95b-9097-4c37-9b91-5bdd8aed3238)

Would be nice to keep the same ordering as in the RegionSelect itself. You could extract the sorting alg from `getRegionOptions` as its own util and use it in both places.

cpathipa replied:
Actually, based on user session feedback, it seems that the order of selected regions should match that of the buckets table, which displays regions in alphabetical order. However, I like the idea of maintaining the same ordering as in RegionSelect and have kept it as the default behavior. This comes with the flexibility of rendering the selected list order through the `sortRegionOptions` prop.

abailly-akamai commented on packages/manager/src/components/RegionSelect/RegionMultiSelect.tsx:
```diff
@@ -0,0 +1,204 @@
+          onChange={(_, selectedOption) =>
+            handleRegionChange(selectedOption as RegionSelectOption[])
+          }
+          renderOption={(props, option, { selected }) => {
```
seconding this, if possible

abailly-akamai commented on packages/manager/src/components/RegionSelect/RegionMultiSelect.tsx:
```diff
@@ -0,0 +1,204 @@
+                  <Box
+                    sx={{
+                      alignItems: 'center',
+                      // display: 'flex',
```
cleanup

abailly-akamai commented on packages/manager/src/components/RegionSelect/RegionSelect.utils.ts:
```diff
@@ -151,3 +152,21 @@ export const getRegionOptionAvailability = ({
+      })
+    )
+    .filter((region): region is RegionSelectOption => !!region);
+};
```
Be consistent with JSDocs comment and making sure you describe your utils

cpathipa replied:
Sure, Added descriptive comments to the utils 135a8dd

abailly-akamai commented on packages/manager/src/components/RegionSelect/RegionSelect.utils.ts:
```diff
@@ -151,3 +152,21 @@ export const getRegionOptionAvailability = ({
+        selectedRegionId,
+      })
+    )
+    .filter((region): region is RegionSelectOption => !!region);
```
can we avoid casting here?

cpathipa replied:
I think we could do something like this 
```}: GetSelectedRegionsByIdsArgs): (RegionSelectOption | undefined)[] => {
  return selectedRegionIds
    .map((selectedRegionId) =>
      getSelectedRegionById({
        accountAvailabilityData,
        currentCapability,
        regions,
        selectedRegionId,
      })
    )
    .filter(Boolean);```

However, this approach might cause us to update all the references to `(RegionSelectOption | undefined)[]`. Also, we may need to refer to it using optional chaining.

abailly-akamai replied:
Why not using a `flatMap`?

```
export const getSelectedRegionsByIds = ({
  accountAvailabilityData,
  currentCapability,
  regions,
  selectedRegionIds,
}: GetSelectedRegionsByIdsArgs): RegionSelectOption[] => {
  return selectedRegionIds.flatMap((selectedRegionId) => {
    const region = getSelectedRegionById({
      accountAvailabilityData,
      currentCapability,
      regions,
      selectedRegionId,
    });
    return region ? [region] : [];
  });
};
```

cpathipa replied:
IMO, the current approach of using `map` followed by `filter` is more straightforward and readable for this specific scenario. It clearly shows the intent of transforming each ID into a region and then filtering out any invalid or non-existent regions. The type guard in the filter (`region` is `RegionSelectOption`) is a clean way to ensure type safety and clarity in the code, making it evident that the final array should only contain valid `RegionSelectOption` objects. Unless there's a specific reason or benefit we are aiming with `flatMap`. I think using `flatMap` directly isn't really suitable in this scenario unless `getSelectedRegionById` return an array.

abailly-akamai replied:
ok. but to be clear, this only provides compile-time type safety, it doesn't guarantee runtime type safety.

cpathipa replied:
Yeah, you are right. Considering the context and requirements, I think we are good with compile-time type safety. Also, we have control over the structure of the data as long as `getSelectedRegionById` returns either a `RegionSelectOption` type object or a falsy value (like null or undefined).

abailly-akamai reviewed (commented):
Thanks for breaking this down in its own component! 

The component tested in Storybook looks good 👍 

Left a couple questions and some comments to improve code 

Also, I am not a fan of this UI pattern in general. When selecting a region, the user has no idea a list is behind the dropdown, and the checkmark next to the region is faint. There's a risk the user does not know the selection worked. I'd like to bring this to UX cause we may want to re-thing this pattern. CC @jaalah-akamai 

![Screenshot 2024-01-22 at 10 28 40](https://github.com/linode/manager/assets/130582365/17b084c3-c209-4c87-a5d7-457494c43c90)

cpathipa commented:
@jaalah-akamai @jaalah-akamai Great feedback.

- Renamed - `RenderSelectedRegionsList`  to `SelectedRegionsList`
- Renamed - `removeOption` to `handleRemoveOption`
- improved readability in story book variable naming - `selectedRegionsIds`, `setSelectedRegionsIds`
- Removed debugger and commented code and described the util function.
- Added UX concern of selected region list and checkmark selection item to our next cafe.
-  Added border under Select All and Deselect All.
- Created a followup ticket M3-7675 for coverage and to make renderOption as reusable for `RegionSelect` and `RegionMultiSelect`.

abailly-akamai commented on packages/manager/src/components/RegionSelect/RegionMultiSelect.tsx:
```diff
@@ -0,0 +1,160 @@
+              : selectedRegions
+          }
+          onRemove={handleRemoveOption}
+        />
```
Still a purely alphabetical order - Not wrong! just wanted to point it out since it differs from the order in the select

![Screenshot 2024-01-23 at 09 06 50](https://github.com/linode/manager/assets/130582365/51e9760b-2b0f-402e-b322-0f982ad6610f)

cpathipa replied:
Sure thing, added a follow-up ticket (M3-7675) to revisit the order of rendering selected regions. At this point, the component supports the flexibility of using a custom component to render the list and sort order.

abailly-akamai commented on packages/manager/src/components/RegionSelect/RegionMultiSelect.stories.tsx:
```diff
@@ -0,0 +1,111 @@
+  onRemove,
+  selectedRegions,
+}: SelectedRegionsProps) => {
+  const handleRemove = (data: RemovableItem) => {
+    onRemove((data as unknown) as RegionSelectOption);
+  };
```
No need to cast here I think

```suggestion
 const handleRemove = (item: RemovableItem) => {
    onRemove(item.data);
 };
```

abailly-akamai reviewed (commented):
Thanks for the changes! - a couple more comments to clean things up and this should be good to go

bnussman-akamai reviewed (approved):
One day, we may want to combine RegionSelect and RegionMultiSelect into one component, but this is good for now!
